File: benchmark/synduce/tree/mips.f
Inductive Tree = nil Unit | node {Int, Tree, Tree};

max = \a: Int. \b: Int. if < a b then b else a;

spec = \t: Tree. (fix (
  \f: {Int, Int} -> Tree -> {Int, Int}. \s: {Int, Int}. \t: Tree.
  match t with
    nil _ -> s
  | node {a, l, r} ->
    let result = (f s l) in
    f {+ a result.1, max result.2 (+ a result.1)} r
  end
) {0,0} t).2;

repr = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    nil _ -> nil unit
  | node {a, l, r} -> node {a, f l, f r}
  end
);

main = \t: Tree. spec (repr t);

File: ml-benchmark/synduce/tree/mips.ml
type tree = Nil of unit | Node of int * tree * tree

let max a b = if a < b then b else a

let spec t =
  let rec f s t =
    match t with
    | Nil _ -> s
    | Node (a, l, r) ->
        let result = f s l in
        match result with
        | (x, y) ->
            let sum = a + x in
            f (sum, max y sum) r
  in
  let res = f (0, 0) t in
  match res with
  | (_, y) -> y

let rec repr t =
  match t with
  | Nil _ -> Nil ()
  | Node (a, l, r) -> Node (a, repr l, repr r)

let program t = spec (repr t)